✨ New WordPress.WP.GetMetaSingle sniff

rodrigoprimo opened the pull request:
This sniff warns when get_\*_meta() and get_metadata\*() functions are used with the $meta_key/$key param, but without the $single parameter ([here](https://github.com/WordPress/WordPress-Coding-Standards/pull/2465/files#diff-41c2b89fbd527d3c67ff73f7b3b477d8a0f1a0dffe904794613d58327c83ad87R39-R46) is the full list of functions that trigger this sniff). This could lead to unexpected behavior as an array will be returned, but a string might be expected.

This PR adds the new sniff to WordPress-Extra as suggested by Juliette in [this Slack thread](https://wordpress.slack.com/archives/C5VCTJGH3/p1718283215160489?thread_ts=1608539112.039700&cid=C5VCTJGH3).

It includes tests and sniff documentation in the XML format.

Related issue: https://github.com/WordPress/WordPress-Coding-Standards/issues/2459

jrfnl commented:
We reviewed this in a pairing session, formal review will follow once the next iteration is up.

rodrigoprimo commented:
@jrfnl, I implemented the changes that you suggested during our pairing session, and this PR is ready for a final review.

rodrigoprimo commented:
When we checked this PR together, you suggested that I change the title of the XML documentation from `Get Meta Single` to `Get Meta Function Single Parameter`. I opted to follow the same pattern and rename the sniff as well from `GetMetaSingle` to `GetMetaFunctionSingleParameter`. Let me know if you prefer the new name or the old name.

jrfnl reviewed (changes requested):
He @rodrigoprimo, thanks for updating the PR!

Mostly looking good. I did leave some remarks inline, mostly small/nitpicky things, but a few to do with precision which really should be fixed.

> When we checked this PR together, you suggested that I change the title of the XML documentation from `Get Meta Single` to `Get Meta Function Single Parameter`. I opted to follow the same pattern and rename the sniff as well from `GetMetaSingle` to `GetMetaFunctionSingleParameter`. Let me know if you prefer the new name or the old name.

I think the new name is very long and  wordy and, like I said when we discussed this before, I think I prefer a name which is in line with prior art in this repo, which largely uses the `FunctionNameParameterNameSniff` pattern, which this sniff now doesn't comply with.

Examples of prior art:
* `CurrentTimeTimestampSniff`
* `PregQuoteDelimiterSniff`
* `StrictInArraySniff` (wrong order, but that's historical, principle is still the same)

rodrigoprimo commented on WordPress/Sniffs/WP/GetMetaFunctionSingleParameterSniff.php:
```diff
@@ -0,0 +1,179 @@
+		$recommended = $this->target_functions[ $matched_content ]['recommended'];
+
+		$meta_key = PassedParameters::getParameterFromStack( $parameters, $condition['position'], $condition['parameter'] );
+		if ( false === $meta_key ) {
+			return;
+		}
+
+		$single = PassedParameters::getParameterFromStack( $parameters, $recommended['position'], $recommended['parameter'] );
+		if ( is_array( $single ) ) {
+			return;
+		}
```
`getParameterFromStack()` returns either an `array` with the param information or `false` if the parameter was not passed. In the first call to this method, the sniff should bail early if the `$[meta_]key` parameter is not passed to the get meta function. Thus the code is checking if `$meta_key` is false. The second call is only executed if the `$[meta_]key` parameter is passed. In this case, the sniff should bail early if the `$single` parameter is also passed. Thus the code is checking if `$single` is an array.

I have a feeling that in the first version that I created, I was using `false !== $single`. But then opted to use `is_array( $single )` as this seems easier to read to me. It gives more information about what the code expects the `$single` variable to be.

Do you think it would be better to use `! is_array( $meta_key )` for the first check?

rodrigoprimo replied:
I should also note that I used `false === $meta_key` instead of `! $meta_key` as the PHPCS coding standards require the former format. I didn't realize that the WPCS coding standards does not require the same format.

jrfnl replied:
Yes, I think checking `! is_array()` make sense here and makes it easier to read/understand the code. :+1:

rodrigoprimo commented on WordPress/Docs/WP/GetMetaFunctionSingleParameterStandard.xml:
```diff
@@ -0,0 +1,34 @@
+    `get_metadata_default()`, and `get_metadata_raw()` functions, if the `$[meta_]key` parameter is
+    passed, the `$single` parameter should be passed as well to make it explicit what is the
+    expected return type.
+    ]]>
```
Good point! I agree that it is worth including a note about this. That being said, unless I'm missing something, I don't think `false` is a valid value for a single meta value. WP converts `false` to an empty string when adding a new meta to the database.

> An empty string if a valid but non-existing comment ID is passed.

Besides that, the documentation for the get meta functions is incorrect as far as I can check. They may return an empty string if a valid but non-existing ID is passed only when `$single` is `true`. If `$single` is `false`, they may return an empty array. I can open Trac ticket to update the documentation.

https://github.com/WordPress/wordpress-develop/blob/4f56c3d18a7236048ed40bfa8c82fb8c88f9b361/src/wp-includes/meta.php#L673-L674

https://github.com/WordPress/wordpress-develop/blob/4f56c3d18a7236048ed40bfa8c82fb8c88f9b361/src/wp-includes/meta.php#L688-L692

Considering the above, I believe we could add the following paragraph:

"Note that for a valid but non-existing ID these functions may return an empty string if `$single` is `true` or an empty array if `$single` is `false`. This can be confusing as these might be valid values for a meta value."

What do you think?

jrfnl replied:
I agree that opening a Trac ticket to fix the return value documentation might be a good idea.

Then again, this could probably be fixed via the catch-all [WP docs Trac ticket for WP 6.7](https://core.trac.wordpress.org/ticket/61608).

Probably simplest to just create a patch, link it to that ticket and then leave a comment on the Trac ticket so people get notified to take a look.

jrfnl replied:
> Good point! I agree that it is worth including a note about this. That being said, unless I'm missing something, I don't think `false` is a valid value for a single meta value. WP converts `false` to an empty string when adding a new meta to the database.

To be honest, that sounds like a bug in WP (or in one of the filters hooked into something called within the `add/update_metadata` functions, as the documentation says that the type of `$value` is `mixed` and doesn't document any restrictions on `mixed` (other than non-scalars being serializable). If the type of the `$value` gets changed in some way, that will lead to lots of WTF situations, so this is problematic.

Would be worth looking into, but not something we need to address in this PR (outside scope).

If recently, it should be fixed as a bug fix.
If this type conversion has been happening for quite a while, fixing it would be a BC-break, so can't be done. In that case, the documentation for `add/update/get_metadata` (and all the variants) should be fixed to document the type casts/juggling being done.

Also would be good to add (probably a sh*tload of) test to document the behaviour and prevent it getting worse in the future ;-)


> Considering the above, I believe we could add the following paragraph:
> 
> "Note that for a valid but non-existing ID these functions may return an empty string if `$single` is `true` or an empty array if `$single` is `false`. This can be confusing as these might be valid values for a meta value."
> 
> What do you think?

Sounds like a good addition.

rodrigoprimo replied:
> I agree that opening a Trac ticket to fix the return value documentation might be a good idea.
> Then again, this could probably be fixed via the catch-all [WP docs Trac ticket for WP 6.7](https://core.trac.wordpress.org/ticket/61608).
> Probably simplest to just create a patch, link it to that ticket and then leave a comment on the Trac ticket so people get notified to take a look.

Here is the comment on the ticket that you mentioned with a link to the PR: https://core.trac.wordpress.org/ticket/61608#comment:9

rodrigoprimo replied:
> If recently, it should be fixed as a bug fix.
> If this type conversion has been happening for quite a while, fixing it would be a BC-break, so can't be done. In that case, the documentation for add/update/get_metadata (and all the variants) should be fixed to document the type casts/juggling being done.

I just created another ticket for this other issue: https://core.trac.wordpress.org/ticket/61950

jrfnl replied:
> Here is the comment on the ticket that you mentioned with a link to the PR: https://core.trac.wordpress.org/ticket/61608#comment:9

That is the automated PR link comment. People don't get notified about those. You need to also leave an actual comment on the Trac ticket to say that a new PR has been added to the ticket (and get the notifications send out).

rodrigoprimo replied:
Thanks, @jrfnl. I was not aware of that. I assumed that an actual comment was not necessary as in that same ticket, other contributors left just the automated PR link comment, and that was enough for some WP core developer to review the PR. I will leave a comment tomorrow.

rodrigoprimo replied:
Thanks! :raised_hands:

rodrigoprimo commented on WordPress/Sniffs/WP/GetMetaFunctionSingleParameterSniff.php:
```diff
@@ -0,0 +1,179 @@
+ * Flags calls to get_(comment|post|site|term|user)_meta(), get_metadata(), get_metadata_default()
+ * and get_metadata_raw() functions that include the $key/$meta_key parameter, but omit the $single
+ * parameter. Omitting $single in this situation can result in unexpected return types and lead to
+ * bugs.
```
Ok, I added a short description before the long description.

rodrigoprimo commented on WordPress/Sniffs/WP/GetMetaFunctionSingleParameterSniff.php:
```diff
@@ -0,0 +1,179 @@
+	protected $target_functions = array(
+		'get_comment_meta'     => array(
+			'condition'   => array(
+				'parameter' => 'key',
```
Yeah, I was not super happy with `parameter`, but I couldn't think of a better name. IMO, `param_name` is a bit better than `name` so I went with it.

rodrigoprimo commented on WordPress/Sniffs/WP/GetMetaFunctionSingleParameterSniff.php:
```diff
@@ -0,0 +1,179 @@
+		$message_data = array( $condition['parameter'], $tokens[ $stackPtr ]['content'], $recommended['parameter'] );
+
+		$this->phpcsFile->addWarning(
+			'When passing the $%s parameter to %s(), it is recommended to pass the $%s parameter as well to make it explicit whether an array or a string is expected.',
```
Good catch, I completely missed that. I tweaked your suggestion a bit. Let me know what you think.

rodrigoprimo commented:
Thanks for your review, @jrfnl. I added new commits addressing all the points that you raised. Could you please take another look?

> I think the new name is very long and wordy and, like I said when we discussed this before, I think I prefer a name which is in line with prior art in this repo, which largely uses the FunctionNameParameterNameSniff pattern, which this sniff now doesn't comply with.

That makes sense. I reverted to using `GetMetaSingle` as the name of the sniff.

jrfnl reviewed (approved):
@rodrigoprimo Thanks for updating the sniff, including the updates made during our video call today in which we did a final review of the sniff.

As discussed, I've now also added a metric to the sniff, just in case someone finds it useful (like when checking the state of things this sniff checks for in plugins in the plugin repo).

As far as I'm concerned, this is ready for merge.

@dingo-d @GaryJones Do you still want to have a look or can I merge this ?

Note: this should be squashed-merged!

dingo-d reviewed (approved):
Looks good to me
